tokio/src/sync/watch.rs

```diff
@@ -138,14 +132,16 @@ struct Shared<T> {
     cancel: AtomicWaker,
 }
 
-#[derive(Debug)]
-struct Watchers {
-    next_id: u64,
-    watchers: FnvHashMap<u64, Arc<WatchInner>>,
-}
+type Watchers = FnvHashSet<Watcher>;
+
+/// The watcher's ID is based on the Arc's pointer.
+#[derive(Clone, Debug)]
+struct Watcher(Arc<WatchInner>);
 
 #[derive(Debug)]
 struct WatchInner {
+    /// Last observed version
+    version: AtomicUsize,
```

Comment:
Yes, memory must still be used. But this moves it from directly on the `Receiver` to the slot we are already allocating for `WatchInner`. This means moving around a `Receiver` is only 2 words instead of 3.

Reply:
Hmm, not moving as much data is a valid improvement. But, I'm still curious about how that stacks up against introducing additional SeqCst ops...

Reply:
The atomic is only accessed from the thread that owns the `Receiver`, correct? In that case, you can just use `Relaxed`. There is no unsafe in this file, so risk is minimal.

Reply:
Personally, I don't really have a strong opinion either way.